src/Pipeline.cpp

```diff
@@ -1105,17 +1105,18 @@ void Pipeline::realize(RealizationArg outputs, const Target &t,
 }
 
 void Pipeline::infer_input_bounds(RealizationArg outputs, const ParamMap &param_map) {
-    Target target = get_jit_target_from_environment();
-
-    compile_jit(target);
+    if (!contents->jit_module.compiled()) {
```

Comment:
I know that this call is used very infrequently, but the implicit read-from-the-env makes me nervous. Could we at least add `const Target &target = Target()` as an argument, as we do everywhere else?

Reply:
(And yes, I realize this is pre-existing behavior, so it's definitely scope creep.)

Reply:
I initially did this, but the thing is that setting the target in the call to realize and infer_input_bounds is sticky - it influences the target that will be used later if one is not specified. I think for people using non-default targets, calling compile_jit(my_target) to set it is clearer.

Reply:
Am I correct that this change here is just a performance optimization? As in it does not change any behavior. If so, I'm fine with not combining any possible change in behavior (or performance regression) with a fix that possibly gets rid of a long standing testing failure.

Reply:
Hmmm... actually I think this does change behavior. Before, infer_input_bounds always returned the bounds for the global JIT target. Now it returns that or whatever ```compile_jit``` was last called with if the Func is already compiled. This is indeed problematic. The previous operation was wrong as evidenced by the bug. We're now changing that to a strangely stateful thing. At the very least I would expect that to require documentation updates. But I do sort of agree that the consistent thing would be to add a target argument.

Another way to go would be to disallow rejitting a Func without some much more explicit mechanism. What we have now is also an invitation to thread issues.

Reply:
The schedule is fixed, so the bounds inferred cannot possibly depend on the target at this point. It's simply cheapest to use already-compiled code.

Reply:
... unless no_bounds_query is set.

Reply:
ok, with my latest push the only user-visible change in behavior (ignoring compile time) is that whereas HL_JIT_TARGET=host-no_bounds_query would previously cause havoc, now it cause rejitting if you do an input bounds query.